Add API and protocol compatibility notes (#1779)
kroxylicious/kroxylicious: docs/assemblies/assembly-overview.adoc

An API provides a convenient means for implementing custom logic within the proxy.

[id='assembly-overview-compatability{context}']
== Compatibility

[id='assembly-overview-compatability-api{context}']
=== APIs

> Kroxylicious follows https://https://semver.org/#semantic-versioning-200[Semantic Versioning] rules. While we are still in the `0.y.z` phase we still take API compatability very seriously, we aim to provide at least two minor releases between deprecation and the removal of that deprecated item.

Reviewer: ```suggestion
Kroxylicious follows https://semver.org/#semantic-versioning-200[Semantic Versioning] rules. While we are still in the `0.y.z` phase we still take API compatability very seriously, we aim to provide at least two minor releases between deprecation and the removal of that deprecated item.
```

> We also consider our configuration model to be a "public API", and thus it will follow the same Semantic Versioning and deprecation rules.

Reviewer 2: Define what you think the "configuration model" is. Does it include the Java API, or are we just promising that we aim to support existing config files which make use of deprecated synatx for the duration of the deprecation period?

Also, because we support plugins, this is not entirely our promise to make. We should make it clear that configuration compatibility also depends on plugin authors.

Author: I was just intending to promise that config files are considered an API, how we parse them is up to us. That said its not purely my call to make, do you think we should include the Java API there as well?

You make a good point about plugin config being outside of our control. I'll clarify that.

Reviewer 2: > do you think we should include the Java API there as well?

No, absolutely not! The trouble is how to phrase what the guarantee is without it being very convoluted. This is painful in part because we don't currently have a first class schema. So you have to read the java, know Jackson well enough to know what YAML it would accept, and then we're saying that any YAML accepted by version X will still be accepted in version Y.

```suggestion
We also consider our configuration file syntax (but not the Java code implementing it) to be a "public API", and thus it will follow the same Semantic Versioning and deprecation rules.
```

> Kubernetes Custom Resources are obviously considered to be a public API, and we will make every effort to evolve them inline with Kubernetes best practice.

Reviewer: ```suggestion
Kubernetes Custom Resources are obviously considered to be a public API, and we will make every effort to evolve them in line with Kubernetes best practice.
```

> Kubernetes Custom Resources are obviously considered to be a public API, and we will make every effort to evolve them in line with Kubernetes best practice.

Reviewer 2: We should link to that those Kube best practices are. We should also be explicit that 

* The CR APIs are currently Alpha and so come with out any compatibility guarantees between releases. 
* the CR APIs are versioned separately (e.g. we might get to 1.0 proxy while still having alpha or beta CR APIs)

Author: > The CR APIs are currently Alpha and so come with out any compatibility guarantees between releases.

We could make the same argument for our API as we are still 0.X but we have chosen not to do so, why would we take a different path with the k8s CR APIs? I was expecting that while Alpha API's allow us to make changes at will we'd follow the same practices we do for the other APIs.

> Kubernetes Custom Resources are obviously considered to be a public API, and we will make every effort to evolve them inline with Kubernetes best practice.

Reviewer: this might be our first mention of anything operator related, it might be a bit odd for users to see a mention of CRs without any futher information about an Operator.

Author: Fair, it just seemed natural to include them given our current focus on building them... What does anyone else think?

Reviewer 3: Might be worth a line in the _Kroxylicious overview_ section before going into compatibility as there's an API sentence there too.

[id='assembly-overview-compatability-wire-protocol{context}']
=== Wire Protocol

> Kroxylicious offers the same backwards & forwards compatability guarantees as Apache Kafka. We will support the same range of client and broker versions as the official Apache Kafka Java client.

Reviewer: Could link off to https://kafka.apache.org/protocol#protocol_compatibility

An API provides a convenient means for implementing custom logic within the proxy.

[id='assembly-overview-compatability{context}']

> == Compatibility

Author: I think its valuable downstream but the versioning is (kind of) different, I've no clear idea how to separate the two.

> == Compatibility

Reviewer 3: If there's a section we want to keep out of downstream, we can use a conditional like Keith did with `include-fortanix-dsm-kms`.  If it's some specific value for versioning, we can use an attribute that we can switch the value for downstream.

[id='assembly-overview-compatability-api{context}']
=== APIs

> Kroxylicious follows https://semver.org/#semantic-versioning-200[Semantic Versioning] rules. While we are still in the `0.y.z` phase we still take API compatability very seriously, we aim to provide at least two minor releases between deprecation and the removal of that deprecated item.

Reviewer 3: ```suggestion
Kroxylicious follows https://semver.org/#semantic-versioning-200[Semantic Versioning] rules. We are committed to API compatibility and aim to provide at least two minor releases between deprecating an item and its removal.
```

> Kroxylicious follows https://semver.org/#semantic-versioning-200[Semantic Versioning] rules. While we are still in the `0.y.z` phase we still take API compatability very seriously, we aim to provide at least two minor releases between deprecation and the removal of that deprecated item.

Reviewer 4: SemVer allows looser compatibility when the major version number is 0. I think it's valuable to note that Kroxylicious is being stricter about API compatibility than the SemVer rules require. 

It would be useful for people unfamiliar with SemVer to explain what's special about major version 0, but I think the reference to the initial development phase should remain.

> Kroxylicious follows https://semver.org/#semantic-versioning-200[Semantic Versioning] rules. While we are still in the initial development (denoted by a major version `0`) phase we still take API compatability very seriously, we aim to provide at least two minor releases between deprecation and the removal of that deprecated item.

Reviewer 3: ```suggestion
Kroxylicious follows https://semver.org/#semantic-versioning-200[Semantic Versioning] rules. While we are still in the initial development phase (denoted by a major version `0`), we still take API compatibility very seriously. We aim to provide at least two minor releases between deprecation and the removal of that deprecated item.
```

We also consider our configuration file syntax a public API (though not the Java model backing it). As such, the syntax follows the same Semantic Versioning and deprecation rules.

Kubernetes custom resources are a public API, and we are making every effort to evolve Kroxylicious custom resources in line with https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definition-versioning/[Kubernetes best practices]. Kubernetes resources have their own versioning scheme, which is independent of the Kroxylicious proxy service version. As a result, we may reach version 1.0.0 of Kroxylicious while still using alpha or beta versions of the custom resources.

==== Third-party plugins

Kroxylicious supports loading third-party plugins to extend the core functionality of the project. While these plugins are configured and loaded as first-class entities within Kroxylicious, we cannot guarantee the compatibility of their APIs or configuration properties.

We do however hold filters and plugins provided by the project to the same standards as the rest of the public API.

[id='assembly-overview-compatibility-wire-protocol{context}']
=== Wire protocol

> Kroxylicious offers the same backwards and forwards compatibility guarantees as https://kafka.apache.org/protocol#protocol_compatibility[Apache Kafka]. We support the same range of client and broker versions as the official Apache Kafka Java client.

Reviewer 5: I don't think it will be obvious to the reader the relationship between the Apache Kafka Java client and this project.   They won't know that Kroxylicious actually depends on the Client codec.

Author: I agree its not obvious, but I'm not sure we want it to be so. 

What I'm trying to say is we adopt the same compatibility rules as upstream Kafka. Maybe the java client is superfluous to the point as we act as both client and broker.
